fix definition of MAX_PATHS frame (#354)
quicwg/multipath: draft-ietf-quic-multipath.md

with MP_RETIRE_CONNECTION_ID frames before adding the newly provided connection ID
to the set of active connection IDs belonging to the specified path.

> Endpoints MUST NOT issue new connection IDs which have Path IDs greater than

Author: ```suggestion
Endpoints MUST NOT issue new connection IDs with Path IDs greater than
```

it SHOULD send a MAX_PATHS frame to inform the peer that it could use larger active
path identifiers.

> 

Author: ```suggestion

```

## MAX_PATHS frames {#max-paths-frame}

> A MAX_PATHS frame (type=0x15228c0b) informs the peer of largest Path ID that
> is permitted to use on any path.

Author: ```suggestion
A MAX_PATHS frame (type=0x15228c0b) informs the peer of the cumulative number of paths
it is permitted to open.
```

~~~
MAX_PATHS Frame {
  Type (i) = 0x15228c0b,

>   Maximum Path Identifier (i),

Author: ```suggestion
  Maximum Paths (i),
```

}
~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_PATHS frames contain the following field:

Maximum Path Identifier:

> : The largest Path ID that can be used for this connection.
>   This value cannot exceed 2^32-1, as it is not
>   secure to use Path IDs larger than 2^32-1 with the security considerations outlined in {{multipath-aead}}.

Author: ```suggestion
: A count of the cumulative number of path that can be opened
  over the lifetime of the connection. This value cannot exceed 2^32-1, as it is not
  secure to use Path IDs larger than 2^32-1 with the security considerations outlined in {{multipath-aead}}.
```

> The Maximum Path Identifier MUST NOT be lower than the value
> advertised in the initial_max_paths transport parameter. Receipt
> of a Path ID lower than the value of initial_max_paths transport parameter
> MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.
> Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
> that uses a Path ID that is larger than the announced Path ID
> MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Author: ```suggestion
The Maximum Paths value MUST NOT be lower than the value
advertised in the initial_max_paths transport parameter. Receipt
of a Maximum Paths value lower than the value of initial_max_paths transport parameter
MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.
Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Maximum Paths value
MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.
```

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with

> a smaller Path ID than was previously received. MAX_PATHS frames that

Author: ```suggestion
a smaller Maximum Paths value than was previously received. MAX_PATHS frames that
```

> do not announce a larger Path ID than previously received MUST be ignored.

Author: ```suggestion
do not announce a larger maximum Paths value than previously received MUST be ignored.
```

Endpoints MUST NOT issue new connection IDs which have path identifiers larger than

> the Path ID announced by the peer in the Maximum Path Identifier field in the MAX_PATHS frame.

Author: ```suggestion
the Maximum Paths value announced by the peer in the MAX_PATHS frame.
```

MAX_PATHS frames contain the following field:

Maximum Path Identifier:

> : A count of the cumulative number of paths that can be opened
>   over the lifetime of the connection. This value MUST NOT exceed 2^32-1, as
>   Path IDs are defined with a maximum value 2^32-1 as the 32 bits of the Path ID are used
>   to calculate the nonce (see Section {{multipath-aead}}).
> 
> The Maximum Paths value MUST NOT be lower than the value
> advertised in the initial_max_paths transport parameter. Receipt
> of a Maximum Paths value lower than the value of initial_max_paths transport parameter
> MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Author: ```suggestion
: A count of the cumulative number of paths that can be opened
  over the lifetime of the connection. This value MUST NOT exceed 2^32-1, as
  Path IDs are defined with a maximum value 2^32-1 as the 32 bits of the Path ID are used
  to calculate the nonce (see Section {{multipath-aead}}).
  The Maximum Paths value MUST NOT be lower than the value
  advertised in the initial_max_paths transport parameter. Receipt
  of an invalid Maximum Paths value MUST be treated as a
  connection error of type MP_PROTOCOL_VIOLATION.
```

  over the lifetime of the connection. This value MUST NOT exceed 2^32-1, as
  Path IDs are defined with a maximum value 2^32-1 as the 32 bits of the Path ID are used
  to calculate the nonce (see Section {{multipath-aead}}).
  The Maximum Paths value MUST NOT be lower than the value
  advertised in the initial_max_paths transport parameter. Receipt
  of an invalid Maximum Paths value MUST be treated as a
  connection error of type MP_PROTOCOL_VIOLATION.

Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Maximum Paths value
MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with
a smaller Maximum Paths value than was previously received. MAX_PATHS frames that
do not increase the path limit MUST be ignored.

Endpoints MUST NOT issue new connection IDs which have path identifiers larger than
the Maximum Paths value announced by the peer in the MAX_PATHS frame.
If no MAX_PATHS frame was received yet, the Maximum Paths value
corresponds to the value of initial_max_paths transport parameter.

> When there are not enough unused path identifiers, endpoints can

Author: ```suggestion
When there are no connection IDs available for unused path identifiers, endpoints can
```

MAX_PATHS frames contain the following field:

Maximum Path Identifier:
: A count of the cumulative number of paths that can be opened
  over the lifetime of the connection. This value MUST NOT exceed 2^32-1, as
  Path IDs are defined with a maximum value 2^32-1 as the 32 bits of the Path ID are used
  to calculate the nonce (see Section {{multipath-aead}}).
  The Maximum Paths value MUST NOT be lower than the value
  advertised in the initial_max_paths transport parameter. Receipt
  of an invalid Maximum Paths value MUST be treated as a
  connection error of type MP_PROTOCOL_VIOLATION.

Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Maximum Paths value
MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with

> a smaller Maximum Paths value than was previously received. MAX_PATHS frames that
> do not announce a larger maximum Paths value than previously received MUST be ignored.

Author: ```suggestion
a smaller Maximum Paths value than was previously received. MAX_PATHS frames that
do not increase the path limit MUST be ignored.
```

Endpoints MUST NOT issue new connection IDs which have path identifiers larger than
the Maximum Paths value announced by the peer in the MAX_PATHS frame.
If no MAX_PATHS frame was received yet, the Maximum Paths value
corresponds to the value of initial_max_paths transport parameter.

> 

Author: ```suggestion
```

## MAX_PATHS frames {#max-paths-frame}

A MAX_PATHS frame (type=0x15228c0b) informs the peer of largest Path ID that
is permitted to use on any path.

MAX_PATHS frames are formatted as shown in {{fig-max-paths-frame-format}}.

~~~
MAX_PATHS Frame {
  Type (i) = 0x15228c0b,
  Maximum Path Identifier (i),
}
~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_PATHS frames contain the following field:

Maximum Path Identifier:
: The largest Path ID that can be used for this connection.
  This value cannot exceed 2^32-1, as it is not

>   possible to encode Path IDs larger than 2^32-1 as variable integer.

Reviewer: ```suggestion
  secure to use Path IDs larger than 2^32-1 with the security considerations outlined in {{multipath-aead}}.
```

Reviewer: @huitema is right. Variable integer is able to encode up to 2^62 -1. 
Using the Path ID as part of the AEAD nonce input limit the upper bound of Path ID.

Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Path ID

> MUST be treated as a connection error of type FRAME_ENCODING_ERROR.

Author: I don't FRAME_ENCODING_ERROR is actually the right error here. It should be MP_PROTOCOL_VIOLATION.

> MUST be treated as a connection error of type FRAME_ENCODING_ERROR.

Author: I don't FRAME_ENCODING_ERROR is actually the right error here. It should be MP_PROTOCOL_VIOLATION.

```suggestion
MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.
```

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with
a smaller Path ID than was previously received. MAX_PATHS frames that
do not announce a larger Path ID than previously received MUST be ignored.

Endpoints SHOULD NOT issue new connection IDs which have path identifiers larger than
the Path ID announced in the Maximum Path Identifier field in the

> MP_MAX_PATHS frame {{max-paths-frame}}.

Author: ```suggestion
MAX_PATHS frame.
```

> If no MAX_PATHS frame was received yet, the maximum Path Identifier

Author: ```suggestion
If no MAX_PATHS frame was received yet, the Maximum Path Identifier
```

corresponds to the value of initial_max_paths transport parameter.

An endpoint MUST NOT initiate a path with a Path ID larger than the Maximum Path Identifier value.
An endpoint MUST terminate the a connection with an error of type MP_PROTOCOL_VIOLATION
if a peer opens a path with a larger Path ID than permitted.

> When there are not enough unused path identifiers, endpoints can
> send a MAX_PATHS frame to inform the peer that new path identifiers are available.

Author: I also proposed to remove this sentence for now. First it is unclear what "not enough" means and second we have two open issues to discuss this further (see #325 and #338).

```suggestion
```

Reviewer: Oh, we need to keep the mechanism for updating the Path ID limit. 
"Not enough" means all the available Path IDs issued by transport parameter and previous MAX_PATHS(_ID) frames are all used in current paths. If endpoints wants to create new paths(also need to issue new CIDs associated with these Path IDs) at this time, they need to use MAX_PATHS(_ID) frame to inform the peer that new used Path ID resources are allowed.

Author: Okay looking at 5.1.1 ([Issuing Connection IDs](https://datatracker.ietf.org/doc/html/rfc9000#name-issuing-connection-ids)) in RFC9000, I think we should not state these rules normatively again, we should just say these rules should now be applied per path.

~~~
MAX_PATHS Frame {
  Type (i) = 0x15228c0b,
  Maximum Path Identifier (i),
}
~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_PATHS frames contain the following field:

Maximum Path Identifier:

> : The latest Path ID that can be used for this connection.

Reviewer 2: Note that the MAX_PATHS frame is a unidirectional advertisement.
```suggestion
: The largest Path ID that can be used for this connection by the receiving endpoint.
```

  This value cannot exceed 2^32-1, as it is not
  possible to encode Path IDs larger than 2^32-1.

Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Path ID
MUST be treated as a connection error of type FRAME_ENCODING_ERROR.

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with
a smaller Path ID than was previously received. MAX_PATHS frames that
do not announce a larger Path ID than previously received MUST be ignored.

Endpoints SHOULD NOT issue new connection IDs which have path identifiers larger than
the Path ID announced in the Maximum Path Identifier field in the

> MP_MAX_PATHS frame {{max-paths-frame}}.

Reviewer 2: I think we should clarify that the MAX_PATHS frame is a unidirectional advertisement to the peer having the same semantic as the `initial_max_paths` TP. To that end, an endpoint MUST NOT issue new CIDs for a Path ID larger than the maximum one advertised by the peer (and not SHOULD NOT) and MUST terminate the a connection with an error of type MP_PROTOCOL_VIOLATION. Allowing (without recommending) an endpoint to issue CIDs with Path ID larger than supported by the receiving host pose security concerns from a resource exhaustion viewpoint.

However, if the peer allows a maximum Path ID X that is larger than the local advertised maximum path ID Y (i.e., X > Y), then an endpoint may (MAY?) provides CIDs for Path IDs Z where Y < Z <= X.

Further, we probably mention the MAX_PATHS frame (and not MP_MAX_PATHS).

Author: But yes, we really need to clarify that the value is unidirectional. I was thinking at some point that this is a global value and thought that if the MAX_PATHS frame gets reordered, a MP_NEW_CONNECTION_ID frame using a new path id could arrive before. However, as MAX_PATHS is sent from one peer and MP_NEW_CONNECTION_ID from the other, this cannot happen. So this is should be a MUST NOT. I will propose further text to clarify tomorrow!

```suggestion
Endpoints MUST NOT issue new connection IDs which have path identifiers larger than
the Path ID announced by the peer in the Maximum Path Identifier field in the
MP_MAX_PATHS frame {{max-paths-frame}}.
```

If no MAX_PATHS frame was received yet, the maximum Path ID

> correspondes to value of initial_max_paths transport parameter.

Reviewer 2: ```suggestion
corresponds to the value of initial_max_paths transport parameter.
```

An endpoint MUST NOT initiate a path with a Path ID larger than the Maximum Path Identifier value.
An endpoint MUST terminate the a connection with an error of type MP_PROTOCOL_VIOLATION

> if a peer opens a path with a larger Path ID than permitted.

Reviewer 2: To me, this whole paragraph can be removed once the corrections of the previous paragraph are integrated.

Author: +1 to remove it

Author: ```suggestion
```

MAX_PATHS frames contain the following field:

Maximum Path Identifier:
: The largest Path ID that can be used for this connection.
  This value cannot exceed 2^32-1, as it is not
  possible to encode Path IDs larger than 2^32-1 as variable integer.

> 

Reviewer 3: Uh, no. We could encode up to 2^62 - 1. The real reason is that we cannot encode more than 32 bit of path-ID and 64 bit of packet number in the 96 bit AEAD nonce.

Author: That doesn't make sense. In this case we would allow 2^62 - 1 path IDs but require a key update earlier.

Reviewer 3: I would be fine with that. It was actually the original design. But there are complexities. For example, can you have both a long-lived path number 1 and then create path ID 2^32 + 1?

Author: hm, okay I guess 2^32 - 1 paths are usually enough... In this case I think we need to actually use a MUST here and we should maybe remove the respective text from the intro in order to define it only once.

It's late here; I will update the PR accordingly tomorrow which gives also others a chance to comment.

Author: Also based on @huitema text below we need to add this:

```suggestion

The Maximum Path Identifier MUST NOT be lower than the value
advertised in the initial_max_paths transport parameter. Receipt
of a Path ID lower than the value of initial_max_paths transport parameter
MUST be treated as a connection error of type MP_PROTOCOL_ERROR.
```

~~~
MAX_PATHS Frame {
  Type (i) = 0x15228c0b,
  Maximum Path Identifier (i),
}
~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_PATHS frames contain the following field:

Maximum Path Identifier:
: The latest Path ID that can be used for this connection.
  This value cannot exceed 2^32-1, as it is not
  possible to encode Path IDs larger than 2^32-1.

Receipt of PATH_AVAILABLE, PATH_STANDBY, PATH_ABANDON or MP_ACK frames
that uses a Path ID that is larger than the announced Path ID
MUST be treated as a connection error of type FRAME_ENCODING_ERROR.

Loss or reordering can cause an endpoint to receive a MAX_PATHS frame with
a smaller Path ID than was previously received. MAX_PATHS frames that
do not announce a larger Path ID than previously received MUST be ignored.

Endpoints SHOULD NOT issue new connection IDs which have path identifiers larger than

> the Path ID announced in the Maximum Path Identifier field in the 

Author: ```suggestion
the Path ID announced in the Maximum Path Identifier field in the
```

MP_MAX_PATHS frame {{max-paths-frame}}.
If no MAX_PATHS frame was received yet, the maximum Path ID
correspondes to value of initial_max_paths transport parameter.

An endpoint MUST NOT initiate a path with a Path ID larger than the Maximum Path Identifier value.
An endpoint MUST terminate the a connection with an error of type MP_PROTOCOL_VIOLATION
if a peer opens a path with a larger Path ID than permitted.

When there are not enough unused path identifiers, endpoints can
send a MAX_PATHS frame to inform the peer that new path identifiers are available.

> 

Reviewer 3: For information, I struggled with the same set of issues when writing the "even numbers" PR. The text that I came up with was:
```
Maximum Paths:
: A count of the cumulative number of paths that can be opened
over the lifetime of the connection.

The endpoint receiving this frame MAY send MP_NEW_CONNECTION_ID frames
with Path IDs strictly lower than twice the Maximum Paths.
For example, if an endpoint received a Maximum Paths value of 3,
it may send MP_NEW_CONNECTION_ID with Path ID 0, 2 and 4, but not higher.

The Maximum Paths value cannot exceed 2^31-1, as it is not
possible to encode Path IDs larger than 2^32-2. Receipt of a
Path Identifier larger than this limit MUST be treated
as a connection error of type FRAME_ENCODING_ERROR.

The Maximum Paths value MUST NOT be lower than the value
advertised in the "initial_max_paths" transport parameter. Receipt
of a value lower than the "initial_max_paths" transport parameter
MUST be treated as a connection error of type MP_PROTOCOL_ERROR.
```
Some of that text was motivated by the odd-even split, but a lot of it was about clarify how the path-ID is used.

Author: We can define the max paths as either the cumulative number of path over the whole life or the largest id. The part that I was trying to fix is that the frame figure said the field is called "Maximum Paths" but the text was defining "Maximum Path Identifier". I thought the later is clearer but in the end we just have to fix it such that it is correct, one or the other way. @huitema do you prefer "Maximum Paths"?

Reviewer 3: MAX_PATHS, MAX_PATH_ID, "bonnet blanc et blanc bonnet", as we say in French. I don't really care.

Author: Okay, but I guess if we use the max path id field, we should also rename the frame... 

Looking at this now again I guess renaming the field in the definition to "Maximum Paths" would have been the smaller chance. I'll revise this PR again tomorrow in order to keep chances as small as possible for now (but ensure things are correct). Sorry for forth and back.
